### ISSUE TITLE
Mysterious slowdown with function using bit shift and bit map

### ISSUE THREAD

**Issue by sesalomon** [GaloisInc/cryptol#537]
```
error : [3][32] -> [3][32]
error [a,b,c] = [c,a^(examplemap c),b^(c >> 1)]
```
replace >> with >>> || <<< || << still long
replace >> with + and goes normal speed 
```
matrix = random 10 : [32][32]

xor : {n} (fin n) => [n] -> Bit
xor xs = ys!0 where ys = [False] # [x^y | x <- xs | y <- ys] 

examplemap : [32] -> [32]
examplemap xs = result
    where result = [xor (xs && row) | row <- matrix]
```
to test issue: 
```
Main> let x = [random 100] # [error xi | xi <- x | i <- [0..100]]
Main> x@30 
```
hitting some cornercase which causes slowdown on an order of magnitude
slight adjustments to "error" function avoid slowdown issue

**Comment by weaversa** [comment]
I can confirm the slowdown on the latest Cryptol.

**Comment by atomb** [comment]
Yes, I can too.

**Comment by weaversa** [comment]
This change makes it immediate:
```
error [a,b,c] = [c,a^(examplemap c),b^(c + 1)]
```

This change still has the slow down:
```
error [a,b,c] = [c,a^(examplemap c),b^c]
```

This is weird.

**Comment by brianhuffman** [comment]
I can tell from the definition of `examplemap` that its return value will be represented as a sequence of individual bits, rather than as a packed word. Also, the `^` and shift/rotates preserve that representation, so it is used throughout. On the other hand, changing `error` to use `+` makes it convert back (at least temporarily) to the packed word representation. So I suspect the problem has to do with how some primitives are implemented for the sequence-of-bits representation. It might be related to #446.

**Comment by brianhuffman** [comment]
I found the problem. It is indeed almost exactly the same problem as #446; I just need to fix a different case in the definition of `wordValLogicOp`.

**Comment by brianhuffman** [comment]
Fixed in 3cee1b0. I looked around to see whether there were any other exponential blowups waiting to be found due to primitives doing the wrong thing with `BitsVal`, but everything looks OK to me now.
